block/produce.go

```diff
@@ -14,72 +14,89 @@ import (
 	tmtime "github.com/tendermint/tendermint/types/time"
 )
 
-// ProduceBlockLoop is calling publishBlock in a loop as long as wer'e synced.
+// ProduceBlockLoop is calling publishBlock in a loop as long as we're synced.
 func (m *Manager) ProduceBlockLoop(ctx context.Context) {
 	m.logger.Debug("Started produce loop")
 
 	ticker := time.NewTicker(m.conf.BlockTime)
 	defer ticker.Stop()
 
-	var tickerEmptyBlocksMaxTime *time.Ticker
-	var tickerEmptyBlocksMaxTimeCh <-chan time.Time
+	// Allow the initial block to be empty
+	produceEmptyBlock := true
+
+	var emptyBlocksTimer <-chan time.Time
+	resetEmptyBlocksTimer := func() {}
 	// Setup ticker for empty blocks if enabled
-	if m.conf.EmptyBlocksMaxTime > 0 {
-		tickerEmptyBlocksMaxTime = time.NewTicker(m.conf.EmptyBlocksMaxTime)
-		tickerEmptyBlocksMaxTimeCh = tickerEmptyBlocksMaxTime.C
-		defer tickerEmptyBlocksMaxTime.Stop()
+	if 0 < m.conf.EmptyBlocksMaxTime {
+		t := time.NewTicker(m.conf.EmptyBlocksMaxTime)
+		emptyBlocksTimer = t.C
+		resetEmptyBlocksTimer = func() {
+			produceEmptyBlock = false
+			t.Reset(m.conf.EmptyBlocksMaxTime)
+		}
+		defer t.Stop()
 	}
 
-	// Allow the initial block to be empty
-	produceEmptyBlock := true
 	for {
 		select {
-		// Context canceled
-		case <-ctx.Done():
+		case <-ctx.Done(): // Context canceled
 			return
-		// If we got a request for an empty block produce it and don't wait for the ticker
-		case <-m.produceEmptyBlockCh:
+		case <-m.produceEmptyBlockCh: // If we got a request for an empty block produce it and don't wait for the ticker
 			produceEmptyBlock = true
-		// Empty blocks timeout
-		case <-tickerEmptyBlocksMaxTimeCh:
-			m.logger.Debug(fmt.Sprintf("No transactions for %.2f seconds, producing empty block", m.conf.EmptyBlocksMaxTime.Seconds()))
+		case <-emptyBlocksTimer: // Empty blocks timeout
 			produceEmptyBlock = true
+			m.logger.Debug(fmt.Sprintf("no transactions, producing empty block: elapsed: %.2f", m.conf.EmptyBlocksMaxTime.Seconds()))
 		// Produce block
 		case <-ticker.C:
-			err := m.produceBlock(ctx, produceEmptyBlock)
-			if errors.Is(err, types.ErrSkippedEmptyBlock) {
+			err := m.produceAndGossipBlock(ctx, produceEmptyBlock)
+			if errors.Is(err, context.Canceled) {
+				m.logger.Error("produce and gossip: context canceled", "error", err)
+				return
+			}
+			if errors.Is(err, ErrRecoverable) {
```

Comment:
In practice, I don't actually think there is anything recoverable inside the produceAndGossip function

Reply:
this check redunedent with the default case

Reply:
good spot